### ISSUE TITLE
TF2 version slower than TF1 version

### ISSUE THREAD

**Issue by rickstaa** [rickstaa/LAC-TF2-TORCH-translation#17]
## User story ##
Since we want to be able to make the PyTorch version as fast as the TF1 version it might be a good idea to start why the TF2 version is slower than the tf1 version.

## Resources
- http://blog.ai.ovgu.de/posts/jens/2019/001_tf20_pitfalls/index.html
- https://www.tensorflow.org/guide/profiler
- https://stackoverflow.com/questions/58441514/why-is-tensorflow-2-much-slower-than-tensorflow-1

## Main takeaways

- Make sure your tf.functions have tensors as their input. If your function accepts Python numeric types each new value actually leads to a new input signature and thus a new graph!!

## TF115

In TF115 `tfp.distributions.ConditionalTransformedDistribution` is replaced by [tfp.distributions.TransformedDistribution](distributions).

## TF2

In TF2 [tfp.bijectors.Affine](https://www.tensorflow.org/probability/api_docs/python/tfp/bijectors/Affine) is replaced by a combination of [tfp.bijector.Shift](https://www.tensorflow.org/probability/api_docs/python/tfp/bijectors/Shift) and [tfp.bijector.Scale](https://www.tensorflow.org/probability/api_docs/python/tfp/bijectors/Scale).

## Possible causes
- [ ] Tf tensor conversion plus graph creation overhead because we input python types to the tf.functions.
- [ ] Gradient computation time.
- [ ] Simulation rollout time
- [ ] Simulation step function
- [ ] Numerical precission during the computation (Input is already float32).
- [ ] Policy sampling.

## Tools
- [spyder4 profiler](https://docs.spyder-ide.org/current/profiler.html)
- [Cprofiler](https://docs.python.org/2/library/profile.html)
- [Tensorflow profiler](https://www.tensorflow.org/tensorboard/tensorboard_profiling_keras)
- [Nvidia profiler if on GPU](https://docs.nvidia.com/cuda/profiler-users-guide/index.html)
- [strace to trace system calls](https://man7.org/linux/man-pages/man1/strace.1.html)
